dart/lib/src/protocol/contexts.dart

```diff
@@ -11,107 +13,170 @@ import 'runtime.dart';
+      this[OperatingSystem.type] = operatingSystem;
+
+  /// Describes a list of runtimes in more detail
+  /// (for instance if you have a Flutter application running
+  /// on top of Android).
+  List<Runtime> get runtimes => List.unmodifiable(this['runtimes']);
+
+  void addRuntime(Runtime runtime) => this['runtimes'].add(runtime);
+
+  void removeRuntime(Runtime runtime) => this['runtimes'].remove(runtime);
 
   /// App context describes the application.
   ///
   /// As opposed to the runtime, this is the actual application that was
   /// running and carries metadata about the current session.
-  final App app;
+  App get app => this[App.type];
+
+  set app(App app) => this[App.type] = app;
 
   /// Carries information about the browser or user agent for web-related
   /// errors.
   ///
   /// This can either be the browser this event ocurred in, or the user
   /// agent of a web request that triggered the event.
-  final Browser browser;
+  Browser get browser => this[Browser.type];
+
+  set browser(Browser browser) => this[Browser.type] = browser;
 
   /// GPU context describes the GPU of the device.
-  final Gpu gpu;
+  Gpu get gpu => this[Gpu.type];
 
-  // TODO: contexts should accept arbitrary values
+  set gpu(Gpu gpu) => this[Gpu.type] = gpu;
 
   /// Produces a [Map] that can be serialized to JSON.
   Map<String, dynamic> toJson() {
     final json = <String, dynamic>{};
 
-    Map<String, dynamic> deviceMap;
-    if (device != null && (deviceMap = device.toJson()).isNotEmpty) {
-      json['device'] = deviceMap;
-    }
-
-    Map<String, dynamic> osMap;
-    if (operatingSystem != null &&
-        (osMap = operatingSystem.toJson()).isNotEmpty) {
-      json['os'] = osMap;
-    }
-
-    Map<String, dynamic> appMap;
-    if (app != null && (appMap = app.toJson()).isNotEmpty) {
-      json['app'] = appMap;
-    }
-
-    Map<String, dynamic> browserMap;
-    if (browser != null && (browserMap = browser.toJson()).isNotEmpty) {
-      json['browser'] = browserMap;
-    }
-
-    Map<String, dynamic> gpuMap;
-    if (gpu != null && (gpuMap = gpu.toJson()).isNotEmpty) {
-      json['gpu'] = gpuMap;
-    }
-
-    if (runtimes != null) {
-      if (runtimes.length == 1) {
-        final runtime = runtimes[0];
-        if (runtime != null) {
-          final key = runtime.key ?? 'runtime';
-          json[key] = runtime.toJson();
-        }
-      } else if (runtimes.length > 1) {
-        for (final runtime in runtimes) {
-          if (runtime != null) {
-            var key = runtime.key ?? runtime.name.toLowerCase();
-
-            if (json.containsKey(key)) {
-              var k = 0;
-              while (json.containsKey(key)) {
-                key = '$key$k';
-                k++;
+    forEach((key, value) {
+      if (value == null) return;
+      switch (key) {
+        case Device.type:
+          Map<String, dynamic> deviceMap;
+          if ((deviceMap = device.toJson()).isNotEmpty) {
+            json['device'] = deviceMap;
```

Comment:
```suggestion
            json[Device.type] = deviceMap;
```

wouldd this work too?

Reply:
✅